### PR TITLE
Task/gh 67 tacc sample plugin

### PR DESCRIPTION
**Pull request by tacc-wbomar**
# To Do

See https://github.com/TACC/Core-CMS/pull/190#issuecomment-821309385.

# Overview

Create a sample TACC plugin (to serve as a bare-bones example plugin).

# Issues

- #67 

# Changes

- _Minor_: Explain how to test TACC CMS Plugins and PostCSS Plugin Configs.
- _Minor_: Support testing without database migrations.
- __New__: Add sample "Greet" plugin.

# Screenshots

<details><summary>Complete Walkthrough of Test Steps (Warning: Many Large Images)</summary>

1 Empty
2 Plugin Search
3 Plugin Config Initial
4 Plugin Add
5 Publish Page
6 Logout
7 Logged Out
8 Return to Page
9 Plugin Output Change
10 Login
11 Plugin Config Change
4 Plugin Add
5 Publish Page
6 Logout
7 Logged Out
8 Return to Page
17 Plugin Output Change
10 Login
19 Plugin Config Change 2
4 Plugin Add
5 Publish Page
6 Logout
7 Logged Out
8 Return to Page
25 Plugin Output Change 2

</details>

# Testing

## Summary

1. Test that plugin greets you by name {{Your_Name}} when:
    - you are logged in.
1. Test that {{Your_Name}} is any of the following:
    - "FirstName LastName"
    - "FirstName"
    - "username"
1. Test that plugin sets custom guest name to "Guest" by default when:
    - you are **not** logged in
1. Test that plugin greets you by a custom guest name "Guest" or "Friend" when:
    - guest name is set
    - you are **not** logged in
1. Test that plugin greets you by default guest name "Guest" when:
    - guest name is **not** set
    - you are **not** logged in
1. Test that unit tests are run and do not fail.
1. Tear down testing (remove plugin instances and uninstall plugin so you can work on other branches).

## Required Setup

1. Enter `cms` cotnainer.
2. Run `python manage.py makemigrations`
3. Run `python manage.py migrate`

## CMS User Testing

<details><summary>Complete Walkthrough of Test Steps (Warning: Long List)</summary>

1. Login.
2. Add plugin to page Content.
3. Use field default value "Guest".
4. Save plugin.
5. __Confirm text says "Hello, {{Your_Name}}".__
6. Publish page.
7. Logout.
8. Log in.
9. Revisit page.
10. __Confirm text says "Hello, Guest".__
11. Login.
12. Edit plugin in page Content.
13. Change field value to "Friend".
14. Save plugin.
15. __Confirm text says "Hello, {{Your_Name}}".__
16. Publish page.
17. Logout.
18. Log in.
19. Revisit page.
20. __Confirm text says "Hello, Friend".__
21. Login.
22. Edit plugin in page Content.
23. Change field value to "" (empty).
24. Save plugin.
25. __Confirm text says "Hello, {{Your_Name}}".__
26. Publish page.
27. Logout.
28. Log in.
29. Revisit page.
30. __Confirm text says "Hello, Guest".__

</details>

## Developer Testing

1. Enter `cms` container.
2. Run `python manage.py test taccsite_cms/contrib/taccsite_sample --nomigrations`.
3. Ensure **all** tests pass.

## Required Teardown

1. Remove plugin instances.
    1. Login as CMS admin.
    2. Delete the instance(s) you added.
1. Uninstall Plugin
    1. Enter `cms` container.
    2. Run `python manage.py cms uninstall plugins TaccsiteSamplePlugin`.

### CODE REVIEW

File: taccsite_cms/contrib/taccsite_sample/tests.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        """Return context necessary for testing plugin logic"""
+        plugin_instance = self.plugin.get_plugin_class_instance()
+        context = plugin_instance.render(self.context, self.plugin, None)
+        return context
+
+    def _get_html(self):
+        """Return markup necessary for testing plugin template"""
+        renderer = ContentRenderer(request=self.factory)
+        html = renderer.render_plugin(self.plugin, self.context)
+        return html
```
Comment (tacc-wbomar) [inline taccsite_cms/contrib/taccsite_sample/tests.py:38]: __What If:__ If tests become standard, then this is boilerplate. But if plugins are moved into their own repo, then sharing this boiler plate may require more overhead.

**Review by tacc-wbomar — COMMENTED** [review]
Notes

**Review by taoteg — CHANGES_REQUESTED** [review]
Everything works as advertised but I ran into two things during testing:

1) We should probably remove `psycopg2==2.8.6` from `requirements.txt` so it won't fail on pip installs (and it is superseded by `psycopg2-binary==2.8.6` on the next line AFAICT).

2) If you are testing outside of a portal (just stand alone CMS) you won't get a username after logging into the cms, instead you get "Hello, ." 

After logging out of the cms I could see the Guest value, and the Guest value changed based on the plugin setting string value set in the CMS admin, so it definitely works.

CAVEAT: Is there any reason (I am missing) to retain the non-binary `psycopg2` plugin at this point? 

```
Error thrown by old psycopg2 plugin:

ollecting psycopg2==2.8.6
  Using cached psycopg2-2.8.6.tar.gz (383 kB)
    ERROR: Command errored out with exit status 1:
     command: /Users/jgentle/Applications/miniconda3/envs/core-portal-v2/bin/python3.6 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-install-g1nt1w6p/psycopg2_03ebb4113f34432d811f04e659c0b589/setup.py'"'"'; __file__='"'"'/private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-install-g1nt1w6p/psycopg2_03ebb4113f34432d811f04e659c0b589/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-pip-egg-info-cujgbtxf
         cwd: /private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-install-g1nt1w6p/psycopg2_03ebb4113f34432d811f04e659c0b589/
    Complete output (23 lines):
    running egg_info
    creating /private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-pip-egg-info-cujgbtxf/psycopg2.egg-info
    writing /private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-pip-egg-info-cujgbtxf/psycopg2.egg-info/PKG-INFO
    writing dependency_links to /private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-pip-egg-info-cujgbtxf/psycopg2.egg-info/dependency_links.txt
    writing top-level names to /private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-pip-egg-info-cujgbtxf/psycopg2.egg-info/top_level.txt
    writing manifest file '/private/var/folders/xp/fbjt5xxn0yb4kvxbrwf6mgf40000gp/T/pip-pip-egg-info-cujgbtxf/psycopg2.egg-info/SOURCES.txt'
    
    Error: pg_config executable not found.
    
    pg_config is required to build psycopg2 from source.  Please add the directory
    containing pg_config to the $PATH or specify the full executable path with the
    option:
    
        python setup.py build_ext --pg-config /path/to/pg_config build ...
    
    or with the pg_config option in 'setup.cfg'.
    
    If you prefer to avoid building psycopg2 from source, please install the PyPI
    'psycopg2-binary' package instead.
    
    For further information please check the 'doc/src/install.rst' file (also at
    <https://www.psycopg.org/docs/install.html>).
    
    ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/fd/ae/98cb7a0cbb1d748ee547b058b14604bd0e9bf285a8e0cc5d148f8a8a952e/psycopg2-2.8.6.tar.gz#sha256=fb23f6c71107c37fd667cb4ea363ddeb936b348bbd6449278eb92c189699f543 (from https://pypi.org/simple/psycopg2/) (requires-python:>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement psycopg2==2.8.6
ERROR: No matching distribution found for psycopg2==2.8.6
```

**Comment by taoteg** [comment]
Aside from my requests, this LGTM!

**Comment by tacc-wbomar** [comment]
1. I did not install Is `psycopg2`. The `…` plugin does not appear to have `psycopg2` as a dependency. Are you sure this is related to this PR?
2. Thank you. I tested on a standalone CMS. Perhaps, it's to do with _LDAP_ENABLED and how the account was set up. I will test this.

**Comment by taoteg** [comment]
IIRC the `psycopg2` is a requirement of the postgresql database for django. We have been using the deprecated package for a while. 

Recently the python gatekeepers got stricter on old packages and now this line: https://github.com/TACC/Core-CMS/blob/2781c7e09cae9f4a9fd2d233bbca03f2966c3ce7/requirements.txt#L74 will fail during a clean pip install. 

It is superseded by the following line: https://github.com/TACC/Core-CMS/blob/2781c7e09cae9f4a9fd2d233bbca03f2966c3ce7/requirements.txt#L75) that installs the currently supported binary version of the exact same package. 

I just thought we could remove the extraneous logging error noise.

In the Core-Portal they use poetry now to manage the dependencies, which has a much better subpackage and deprecation handling workflow. We could consider switching to poetry for the CMS too (in a separate ticket of course).

**Comment by tacc-wbomar** [comment]
Okay, thank you.

As agreed, we'll do a separate issue/PR for the `psycopg2` deprecation.

And I realized the issue is that your user has no first name nor last name.

So:
- [ ] I'll make sure the username is a fallback value.
- [ ] You can extract the `psycopg2` deprecation (possibly as "move to Poetry") to a new issue/PR.

**Comment by tacc-wbomar** [comment]
@jgentle,

I have completed the todo's.

- new ticket https://github.com/TACC/Core-CMS/issues/198 ("If Safe, Then Deprecate psycopg2")
- code now output user's name as "FirstName LastName", "FirstName", or "username"
    - [`…/models.py`: `get_name`](https://github.com/TACC/Core-CMS/pull/190/files#diff-652229134da1b9891b1d0c66ce0bc0138ed8781cbfcb088be9d004dea95f8b73R42-R45)
    - [`…/utils.py`: `has_proper_name`](https://github.com/TACC/Core-CMS/pull/190/files#diff-fdbe43eac0b6db05d026d3ba48b499c67ca32c9824156a46acba1ce1c9a5c664R6)
    - [`…/utils.py`: `get_proper_name`](https://github.com/TACC/Core-CMS/pull/190/files#diff-fdbe43eac0b6db05d026d3ba48b499c67ca32c9824156a46acba1ce1c9a5c664R27)

P.S. Unit tests now also test custom/default guest name and creating user's name based on available data.

**Pull request merged**